File: src/js/AdvancedSearch.js
/**
*  advanced-search
*  Transform any search box / search term query into an advanced multi-field search with custom search operators (PHP, Node/XPCOM/JS, Python)
*
*  @version 0.2.0
*  https://github.com/foo123/advanced-search
**/
!function( root, name, factory ) {
"use strict";
var m;
if ( ('undefined'!==typeof Components)&&('object'===typeof Components.classes)&&('object'===typeof Components.classesByID)&&Components.utils&&('function'===typeof Components.utils['import']) ) /* XPCOM */
    (root.EXPORTED_SYMBOLS = [ name ]) && (root[ name ] = factory.call( root ));
else if ( ('object'===typeof module)&&module.exports ) /* CommonJS */
    module.exports = factory.call( root );
else if ( ('function'===typeof(define))&&define.amd&&('function'===typeof(require))&&('function'===typeof(require.specified))&&require.specified(name) ) /* AMD */
    define(name,['require','exports','module'],function( ){return factory.call( root );});
else if ( !(name in root) ) /* Browser/WebWorker/.. */
    (root[ name ] = (m=factory.call( root )))&&('function'===typeof(define))&&define.amd&&define(function( ){return m;} );
}(  /* current root */          this, 
    /* module name */           "AdvancedSearch",
    /* module factory */        function( exports, undef ) {
"use strict";

var PROTO = 'prototype', HAS = 'hasOwnProperty', 
    toString = Object[PROTO].toString,
    trim_re = /^\s+|\s+$/g,
    trim = String[PROTO].trim
        ? function( s ){ return s.trim(); }
        : function( s ){ return s.replace(trim_re, ''); },
    AdvancedSearch
;

AdvancedSearch = function AdvancedSearch( operators, aliases ) {
    var self = this;
    if ( !(self instanceof AdvancedSearch) ) return new AdvancedSearch(operators, aliases);
    self.operators = !operators ? ['>=','<=','<','>','~','='] : operators;
    self.aliases = aliases || [];
    self.q = null;
};
AdvancedSearch.VERSION = '0.2.0';
AdvancedSearch.parse = function parse( query, operators, aliases, delims ) {
    // parse advanced search
    var factors = [],
        or_delim, and_delim, or_delim_len, and_delim_len, op_len,
        l, i, c, j, in_string, terms, term, field, op
    ;
    
    query = trim('' + query);
    if ( !query || !query.length ) return factors;
    
    aliases = aliases || {};
    delims = delims || {};
    or_delim = delims[HAS]('or') ? delims['or'] : ',';
    and_delim = delims[HAS]('and') ? delims['and'] : ' ';
    or_delim_len = or_delim.length;
    and_delim_len = and_delim.length;
    op_len = operators.length;
    
    l = query.length; i = 0; in_string = null;
    terms = []; field = null; op = null; term = '';
    while ( i < l )
    {
        c = query.charAt( i );
        if ( '"' === c || "'" === c )
        {
            if ( in_string === c )
            {
                in_string = null;
            }
            else if ( in_string )
            {
                term += c;
            }
            else
            {
                in_string = c;
            }
            i++;
            continue;
        }
        else if ( in_string )
        {
            term += c;
            i++;
            continue;
        }
        
        if ( null == op )
        {
            for(j=0; j<op_len; j++)
            {
                if ( operators[j] === query.slice( i,i+operators[j].length ) )
                {
                    op = operators[j];
                    if ( !field && term.length )
                    {
                        field = term;
                        term = '';
                    }
                    if ( field && aliases[HAS](field) ) field = aliases[field];
                    break;
                }
            }
            if ( op )
            {
                i += op.length;
                continue;
            }
        }
        
        if ( or_delim === query.slice( i,i+or_delim_len ) )
        {
            if ( term.length ) terms.push( {field:field, op:op, term:term} );
            if ( terms.length ) factors.push( terms );
            terms = [];
            field = null; op = null; term = '';
            i += or_delim_len;
        }
        else if ( and_delim === query.slice( i,i+and_delim_len ) )
        {
            if ( term.length ) terms.push( {field:field, op:op, term:term} );
            field = null; op = null; term = '';
            i += and_delim_len;
        }
        else
        {
            term += c;
            i++;
        }
    }
    if ( term.length ) terms.push( {field:field, op:op, term:term} );
    if ( terms.length ) factors.push( terms );
    return factors;
};
AdvancedSearch[PROTO] = {
    constructor: AdvancedSearch
    
    ,q: null
    ,factors: null
    ,operators: null
    ,aliases: null
    
    ,dispose: function( ) {
        var self = this;
        self.q = null;
        self.factors = null;
        self.operators = null;
        self.aliases = null;
        return self;
    }
    
    ,query: function( query, delims ) {
        var self = this;
        self.q = query;
        return self.factors = AdvancedSearch.parse( query, self.operators, self.aliases, delims||{} );
    }
};
// export it
return AdvancedSearch;
})

File: src/python/AdvancedSearch.py
# -*- coding: UTF-8 -*-
##
#  advanced-search
#  Transform any search box / search term query into an advanced multi-field search with custom search operators (PHP, Node/XPCOM/JS, Python)
#
#  @version 0.2.0
#  https://github.com/foo123/advanced-search
##

class AdvancedSearch:
    """
    Transform any search box / search term query into an advanced multi-field search with custom search operators (PHP, Node/XPCOM/JS, Python)
    https://github.com/foo123/advanced-search
    """

    VERSION = '0.2.0'
    
    def parse( query, operators, aliases=None, delims=None ):
        # parse advanced search
        factors = []
        
        query = ('' + query).strip()
        if not query or not len(query): return factors
        
        if not aliases: aliases = {}
        if not delims: delims = {}
        or_delim = delims['or'] if 'or' in delims else ','
        and_delim = delims['and'] if 'and' in delims else ' '
        or_delim_len = len(or_delim)
        and_delim_len = len(and_delim)
        op_len = len(operators)
        
        l = len(query)
        i = 0
        in_string = None
        terms = []
        field = None
        op = None
        term = ''
        while i < l:
            c = query[ i ]
            if '"' == c or "'" == c:
                if in_string == c:
                    in_string = None
                elif in_string:
                    term += c
                else:
                    in_string = c
                i += 1
                continue
            elif in_string:
                term += c
                i += 1
                continue
            
            if not op:
                for oper in operators:
                    if oper == query[ i:i+len(oper) ]:
                        op = oper
                        if not field and len(term):
                            field = term
                            term = ''
                        if field and (field in aliases): field = aliases[field]
                        break
                if op:
                    i += len(op)
                    continue
            
            if or_delim == query[ i:i+or_delim_len ]:
                if len(term): terms.append( {'field':field, 'op':op, 'term':term} )
                if len(terms): factors.append( terms )
                terms = []
                field = None
                op = None
                term = ''
                i += or_delim_len
            elif and_delim == query[ i:i+and_delim_len ]:
                if len(term): terms.append( {'field':field, 'op':op, 'term':term} )
                field = None
                op = None
                term = ''
                i += and_delim_len
            else:
                term += c
                i += 1
        
        if len(term): terms.append( {'field':field, 'op':op, 'term':term} )
        if len(terms): factors.append( terms )
        return factors
    
    def __init__( self, operators=None, aliases=None ):
        self.operators = ['>=','<=','<','>','~','='] if not operators else operators
        self.aliases = {} if not aliases else aliases
        self.q = None
        self.factors = None
    
    def dispose( self ):
        self.q = None
        self.factors = None
        self.operators = None
        self.aliases = None
        return self
    
    def query( self, query, delims=None ):
        self.q = query
        self.factors = AdvancedSearch.parse( query, self.operators, self.aliases, delims )
        return self.factors



# if used with 'import *'
__all__ = ['AdvancedSearch']
        